File: hashmaps/degree_of_an_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_shortest_sub_array(nums)
    freq = {}
    first = {}
    last = {}
    
    nums.each_with_index do |n, i|
        freq[n] = (freq[n] || 0) + 1
        first[n] = i if first[n].nil?
        last[n] = i
    end
    
    degree = freq.values.max
    
    freq.select { |n, f| f == degree }
        .map { |n, _| last[n] - first[n] + 1 }
        .min
end

File: hashmaps/degree_of_an_array.py
class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        freq = {}
        first = {}
        last = {}

        for i, n in enumerate(nums):
            freq[n] = freq.get(n, 0) + 1
            if n not in first:
                first[n] = i
            last[n] = i

        degree = max(freq.values())

        min_length = len(nums)
        for n in freq:
            if freq[n] == degree:
                min_length = min(min_length, last[n] - first[n] + 1)

        return min_length
            